elastic/kibana
# [Reporting] Create reports with full state required to generate the report

### Summary

#### New export types

* `printable_pdf_v2`
* `png_v2`

The general strategy was to copy as much as possible from the existing PDF and PNG export types altering only how the existing export types handle their URLs. Specifically, we want to use locators to power forwarding state and navigation.

#### Reporting client-side redirect app

This contribution also introduces a new Reporting redirect client app that will be used by the new export types to load app state without using URLs. The strategy is as follows:

1. Client app requests a new report of type PNG/PDF v2 and provides locator params containing the client's locator ID (i.e., clients need a locator implementation), the state _version_ (which will be used to migrate locator state) and the actual state (contained inside of `params`). Collectively these are the `locatorParams` and they are stored in an array so that multiple pages can be included in single report.
2. Reporting creates a new job saved object that will contain the `locatorParams`.
3. When the report job executes, the job saved object will be loaded and a URL + `locatorParam` combination will be created for each page that needs to be included in the report.
4. Each of the URLs is navigated to at which point the redirect app will read the job `locatorParams` injected by chromium and use each of these to retrieve a `locator`
5. Navigation is performed  the `share` plugin which also migrates the app state

#### This contribution does not contain

The following should be addressed in a follow up PR(s):

* A way to pass on `forceNow` to apps that are generating vis. It is not clear that Reporting should need to do this in practice, the plugin rendering the UI should be able to determine this via the locator state or by detecting that they are in screenshot mode (or a combination of those). Open to other input here though!
* Documentation about how to integrate, although the reporting example app was updated to use both the PDF and PNG v2 export types.
* Test coverage for the new export types
* Improved redirect error handling

### To reviewers

To test this functionality:

1. Start kibana (with `yarn start --run-examples`) and ES on basic
2. Go to the "Developer examples" plugins
3. Go to Reporting integration
4. Generate a PDF/PNG

You should get a report that looks like:

![Screenshot 2021-07-22 at 15 28 41](https://user-images.githubusercontent.com/8155004/126646892-084e8c51-7a73-44fb-8393-30825d184a5f.png)

* note the additional forwarded state and the `migrated` value that gets added from running the locator based migrations
* note the `forceNow` value that is added to the locator params. Current thinking is that consumer locators can use this to forward to the place we are navigating to control the "now" of when the report is being generated.



#### Notes on the current implementation

We wanted to preserve all state stored on report/job saved object. This meant using the short URL service is slightly less desirable (but possible).

We also did not want to rewrite the share redirect endpoint to achieve this functionality.

### Screenshots

#### Error state in redirect app

(This particular error is out-of-date with the current code, but the presentation of errors is still the same).

![Screenshot 2021-07-15 at 11 31 37](https://user-images.githubusercontent.com/8155004/125814039-a268b5e6-6e2f-4a7a-9393-2c8d4883ede6.png)

### Future work

* Tests! It would be great to add a test that checks whether a given PDF matches a snapshot PDF (as suggested by @tsullivan )
* Improved error reporting from the redirect process
* A nice enhancement would be for the new v2 reports to link users from the Reporting UI to the reporting location in Kibana (with full app state forwarded)
* Write docs for how to integrate with the new v2 PDF/PNG report types
* Update integration points (like dashboard)


### Checklist
